File: backend/app/application/schemas/dms.py
"""
Módulo DMS (Document Management System) — Schemas Pydantic
"""
from __future__ import annotations
from datetime import datetime
from typing import Optional, List
from decimal import Decimal
from pydantic import BaseModel, ConfigDict


# ─── DMSCarpeta ────────────────────────────────────────────────────────────────

class DMSCarpetaCreate(BaseModel):
    nombre: str
    descripcion: Optional[str] = None
    padre_id: Optional[int] = None
    ruta: Optional[str] = None
    icono: Optional[str] = None
    color: Optional[str] = None
    es_publica: bool = False
    creado_por_id: Optional[int] = None

class DMSCarpetaUpdate(BaseModel):
    nombre: Optional[str] = None
    descripcion: Optional[str] = None
    padre_id: Optional[int] = None
    ruta: Optional[str] = None
    icono: Optional[str] = None
    color: Optional[str] = None
    es_publica: Optional[bool] = None

class DMSCarpetaResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    nombre: str
    descripcion: Optional[str] = None
    padre_id: Optional[int] = None
    ruta: Optional[str] = None
    icono: Optional[str] = None
    color: Optional[str] = None
    es_publica: bool
    creado_por_id: Optional[int] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


# ─── DMSCategoria ──────────────────────────────────────────────────────────────

class DMSCategoriaCreate(BaseModel):
    nombre: str
    descripcion: Optional[str] = None
    codigo: Optional[str] = None
    icono: Optional[str] = None
    color: Optional[str] = None
    activo: bool = True

class DMSCategoriaUpdate(BaseModel):
    nombre: Optional[str] = None
    descripcion: Optional[str] = None
    codigo: Optional[str] = None
    icono: Optional[str] = None
    color: Optional[str] = None
    activo: Optional[bool] = None

class DMSCategoriaResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    nombre: str
    descripcion: Optional[str] = None
    codigo: Optional[str] = None
    icono: Optional[str] = None
    color: Optional[str] = None
    activo: bool
    created_at: Optional[datetime] = None


# ─── DMSTipoDocumento ──────────────────────────────────────────────────────────

class DMSTipoDocumentoCreate(BaseModel):
    nombre: str
    descripcion: Optional[str] = None
    categoria_id: int
    codigo: Optional[str] = None
    extensiones_permitidas: Optional[str] = None
    requiere_firma: bool = False
    requiere_aprobacion: bool = False
    dias_vigencia: Optional[int] = None
    activo: bool = True

class DMSTipoDocumentoUpdate(BaseModel):
    nombre: Optional[str] = None
    descripcion: Optional[str] = None
    categoria_id: Optional[int] = None
    codigo: Optional[str] = None
    extensiones_permitidas: Optional[str] = None
    requiere_firma: Optional[bool] = None
    requiere_aprobacion: Optional[bool] = None
    dias_vigencia: Optional[int] = None
    activo: Optional[bool] = None

class DMSTipoDocumentoResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    nombre: str
    descripcion: Optional[str] = None
    categoria_id: int
    codigo: Optional[str] = None
    extensiones_permitidas: Optional[str] = None
    requiere_firma: bool
    requiere_aprobacion: bool
    dias_vigencia: Optional[int] = None
    activo: bool
    created_at: Optional[datetime] = None


# ─── DMSCampoMetadato ──────────────────────────────────────────────────────────

class DMSCampoMetadatoCreate(BaseModel):
    tipo_documento_id: int
    nombre: str
    etiqueta: str
    tipo_dato: str = "texto"
    opciones: Optional[str] = None
    requerido: bool = False
    orden: int = 0

class DMSCampoMetadatoUpdate(BaseModel):
    nombre: Optional[str] = None
    etiqueta: Optional[str] = None
    tipo_dato: Optional[str] = None
    opciones: Optional[str] = None
    requerido: Optional[bool] = None
    orden: Optional[int] = None

class DMSCampoMetadatoResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    tipo_documento_id: int
    nombre: str
    etiqueta: str
    tipo_dato: str
    opciones: Optional[str] = None
    requerido: bool
    orden: int
    created_at: Optional[datetime] = None


# ─── DMSDocumento ──────────────────────────────────────────────────────────────

class DMSDocumentoCreate(BaseModel):
    codigo: Optional[str] = None
    nombre: str
    descripcion: Optional[str] = None
    tipo_documento_id: Optional[int] = None
    carpeta_id: Optional[int] = None
    estado: str = "BORRADOR"
    version_actual: str = "1.0"
    version_numero: int = 1
    tags: Optional[str] = None
    es_confidencial: bool = False
    permite_descarga: bool = True
    permite_impresion: bool = True
    fecha_vigencia_inicio: Optional[datetime] = None
    fecha_vigencia_fin: Optional[datetime] = None
    propietario_id: Optional[int] = None
    hcm_colaborador_id: Optional[int] = None
    hcm_conductor_id: Optional[int] = None
    tms_vehiculo_id: Optional[int] = None
    modulo_origen: Optional[str] = None
    referencia_externa_id: Optional[int] = None

class DMSDocumentoUpdate(BaseModel):
    codigo: Optional[str] = None
    nombre: Optional[str] = None
    descripcion: Optional[str] = None
    tipo_documento_id: Optional[int] = None
    carpeta_id: Optional[int] = None
    estado: Optional[str] = None
    version_actual: Optional[str] = None
    version_numero: Optional[int] = None
    tags: Optional[str] = None
    es_confidencial: Optional[bool] = None
    permite_descarga: Optional[bool] = None
    permite_impresion: Optional[bool] = None
    fecha_vigencia_inicio: Optional[datetime] = None
    fecha_vigencia_fin: Optional[datetime] = None
    propietario_id: Optional[int] = None
    hcm_colaborador_id: Optional[int] = None
    hcm_conductor_id: Optional[int] = None
    tms_vehiculo_id: Optional[int] = None
    modulo_origen: Optional[str] = None
    referencia_externa_id: Optional[int] = None

class DMSDocumentoResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    codigo: Optional[str] = None
    nombre: str
    descripcion: Optional[str] = None
    tipo_documento_id: Optional[int] = None
    carpeta_id: Optional[int] = None
    estado: str
    version_actual: str
    version_numero: int
    tags: Optional[str] = None
    es_confidencial: bool
    permite_descarga: bool
    permite_impresion: bool
    fecha_vigencia_inicio: Optional[datetime] = None
    fecha_vigencia_fin: Optional[datetime] = None
    propietario_id: Optional[int] = None
    hcm_colaborador_id: Optional[int] = None
    hcm_conductor_id: Optional[int] = None
    tms_vehiculo_id: Optional[int] = None
    modulo_origen: Optional[str] = None
    referencia_externa_id: Optional[int] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

class DMSDocumentoListResponse(BaseModel):
    """Schema reducido para listados paginados de documentos."""
    model_config = ConfigDict(from_attributes=True)
    id: int
    codigo: Optional[str] = None
    nombre: str
    tipo_nombre: Optional[str] = None
    estado: str
    version_actual: str
    fecha_vigencia_inicio: Optional[datetime] = None
    fecha_vigencia_fin: Optional[datetime] = None
    propietario_nombre: Optional[str] = None
    created_at: Optional[datetime] = None


# ─── DMSVersion ────────────────────────────────────────────────────────────────

class DMSVersionCreate(BaseModel):
    documento_id: int
    numero_version: str
    version_numero: int
    es_mayor: bool = False
    nombre_archivo: str
    ruta_archivo: Optional[str] = None
    url_archivo: Optional[str] = None
    tamanio_bytes: Optional[int] = None
    tipo_mime: Optional[str] = None
    hash_md5: Optional[str] = None
    comentario: Optional[str] = None
    ocr_texto: Optional[str] = None
    creado_por_id: Optional[int] = None

class DMSVersionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    documento_id: int
    numero_version: str
    version_numero: int
    es_mayor: bool
    nombre_archivo: str
    ruta_archivo: Optional[str] = None
    url_archivo: Optional[str] = None
    tamanio_bytes: Optional[int] = None
    tipo_mime: Optional[str] = None
    hash_md5: Optional[str] = None
    comentario: Optional[str] = None
    ocr_texto: Optional[str] = None
    creado_por_id: Optional[int] = None
    created_at: Optional[datetime] = None


# ─── DMSMetadatoValor ──────────────────────────────────────────────────────────

class DMSMetadatoValorCreate(BaseModel):
    documento_id: int
    campo_id: int
    valor_texto: Optional[str] = None
    valor_numero: Optional[Decimal] = None
    valor_fecha: Optional[datetime] = None
    valor_booleano: Optional[bool] = None

class DMSMetadatoValorResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    documento_id: int
    campo_id: int
    valor_texto: Optional[str] = None
    valor_numero: Optional[Decimal] = None
    valor_fecha: Optional[datetime] = None
    valor_booleano: Optional[bool] = None
    created_at: Optional[datetime] = None


# ─── DMSFirma ──────────────────────────────────────────────────────────────────

class DMSFirmaCreate(BaseModel):
    documento_id: int
    version_id: Optional[int] = None
    firmante_id: int
    tipo_firma: str
    estado: str = "PENDIENTE"
    fecha_firma: Optional[datetime] = None
    ip_firma: Optional[str] = None
    dispositivo: Optional[str] = None
    observaciones: Optional[str] = None
    orden: int = 0

class DMSFirmaUpdate(BaseModel):
    estado: Optional[str] = None
    fecha_firma: Optional[datetime] = None
    ip_firma: Optional[str] = None
    dispositivo: Optional[str] = None
    observaciones: Optional[str] = None

class DMSFirmaResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    documento_id: int
    version_id: Optional[int] = None
    firmante_id: int
    tipo_firma: str
    estado: str
    fecha_firma: Optional[datetime] = None
    ip_firma: Optional[str] = None
    dispositivo: Optional[str] = None
    observaciones: Optional[str] = None
    orden: int
    created_at: Optional[datetime] = None


# ─── DMSWorkflow ───────────────────────────────────────────────────────────────

class DMSWorkflowCreate(BaseModel):
    nombre: str
    descripcion: Optional[str] = None
    tipo_documento_id: Optional[int] = None
    activo: bool = True
    dias_limite: Optional[int] = None

class DMSWorkflowUpdate(BaseModel):
    nombre: Optional[str] = None
    descripcion: Optional[str] = None
    tipo_documento_id: Optional[int] = None
    activo: Optional[bool] = None
    dias_limite: Optional[int] = None

class DMSWorkflowResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    nombre: str
    descripcion: Optional[str] = None
    tipo_documento_id: Optional[int] = None
    activo: bool
    dias_limite: Optional[int] = None
    created_at: Optional[datetime] = None


# ─── DMSWorkflowPaso ───────────────────────────────────────────────────────────

class DMSWorkflowPasoCreate(BaseModel):
    workflow_id: int
    nombre: str
    tipo: str
    orden: int = 0
    responsable_id: Optional[int] = None
    responsable_rol: Optional[str] = None
    dias_limite: Optional[int] = None
    es_obligatorio: bool = True

class DMSWorkflowPasoUpdate(BaseModel):
    nombre: Optional[str] = None
    tipo: Optional[str] = None
    orden: Optional[int] = None
    responsable_id: Optional[int] = None
    responsable_rol: Optional[str] = None
    dias_limite: Optional[int] = None
    es_obligatorio: Optional[bool] = None

class DMSWorkflowPasoResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    workflow_id: int
    nombre: str
    tipo: str
    orden: int
    responsable_id: Optional[int] = None
    responsable_rol: Optional[str] = None
    dias_limite: Optional[int] = None
    es_obligatorio: bool
    created_at: Optional[datetime] = None


# ─── DMSInstancia ──────────────────────────────────────────────────────────────

class DMSInstanciaCreate(BaseModel):
    workflow_id: int
    documento_id: int
    iniciado_por_id: Optional[int] = None
    paso_actual: int = 0
    fecha_inicio: Optional[datetime] = None
    fecha_limite: Optional[datetime] = None

class DMSInstanciaResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    workflow_id: int
    documento_id: int
    estado: str
    iniciado_por_id: Optional[int] = None
    paso_actual: int
    fecha_inicio: Optional[datetime] = None
    fecha_limite: Optional[datetime] = None
    fecha_fin: Optional[datetime] = None
    comentario_cierre: Optional[str] = None
    created_at: Optional[datetime] = None


# ─── DMSInstanciaPaso ──────────────────────────────────────────────────────────

class DMSInstanciaPasoUpdate(BaseModel):
    estado: Optional[str] = None
    asignado_a_id: Optional[int] = None
    fecha_asignacion: Optional[datetime] = None
    fecha_respuesta: Optional[datetime] = None
    comentario: Optional[str] = None
    accion: Optional[str] = None

class DMSInstanciaPasoResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    instancia_id: int
    paso_id: int
    estado: str
    asignado_a_id: Optional[int] = None
    fecha_asignacion: Optional[datetime] = None
    fecha_respuesta: Optional[datetime] = None
    comentario: Optional[str] = None
    accion: Optional[str] = None
    created_at: Optional[datetime] = None


# ─── DMSExpediente ─────────────────────────────────────────────────────────────

class DMSExpedienteCreate(BaseModel):
    codigo: Optional[str] = None
    nombre: str
    tipo: str
    estado: str = "activo"
    descripcion: Optional[str] = None
    hcm_colaborador_id: Optional[int] = None
    hcm_conductor_id: Optional[int] = None
    tms_vehiculo_id: Optional[int] = None
    propietario_id: Optional[int] = None
    completitud_pct: Optional[Decimal] = None

class DMSExpedienteUpdate(BaseModel):
    codigo: Optional[str] = None
    nombre: Optional[str] = None
    tipo: Optional[str] = None
    estado: Optional[str] = None
    descripcion: Optional[str] = None
    hcm_colaborador_id: Optional[int] = None
    hcm_conductor_id: Optional[int] = None
    tms_vehiculo_id: Optional[int] = None
    propietario_id: Optional[int] = None
    completitud_pct: Optional[Decimal] = None

class DMSExpedienteResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    codigo: Optional[str] = None
    nombre: str
    tipo: str
    estado: str
    descripcion: Optional[str] = None
    hcm_colaborador_id: Optional[int] = None
    hcm_conductor_id: Optional[int] = None
    tms_vehiculo_id: Optional[int] = None
    propietario_id: Optional[int] = None
    completitud_pct: Optional[Decimal] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


# ─── DMSExpedienteDocumento ────────────────────────────────────────────────────

class DMSExpedienteDocumentoCreate(BaseModel):
    expediente_id: int
    documento_id: int
    tipo_requerido: Optional[str] = None
    es_obligatorio: bool = False

class DMSExpedienteDocumentoResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    expediente_id: int
    documento_id: int
    tipo_requerido: Optional[str] = None
    es_obligatorio: bool
    created_at: Optional[datetime] = None


# ─── DMSRetencion ─────────────────────────────────────────────────────────────

class DMSRetencionCreate(BaseModel):
    nombre: str
    tipo_documento_id: Optional[int] = None
    dias_retencion_activo: int
    dias_retencion_total: int
    accion_vencimiento: str = "archivar"
    normativa: Optional[str] = None
    activo: bool = True

class DMSRetencionUpdate(BaseModel):
    nombre: Optional[str] = None
    tipo_documento_id: Optional[int] = None
    dias_retencion_activo: Optional[int] = None
    dias_retencion_total: Optional[int] = None
    accion_vencimiento: Optional[str] = None
    normativa: Optional[str] = None
    activo: Optional[bool] = None

class DMSRetencionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    nombre: str
    tipo_documento_id: Optional[int] = None
    dias_retencion_activo: int
    dias_retencion_total: int
    accion_vencimiento: str
    normativa: Optional[str] = None
    activo: bool
    created_at: Optional[datetime] = None


# ─── DMSAuditoria ─────────────────────────────────────────────────────────────

class DMSAuditoriaResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    documento_id: Optional[int] = None
    version_id: Optional[int] = None
    usuario_id: Optional[int] = None
    accion: str
    detalle: Optional[str] = None
    ip_origen: Optional[str] = None
    user_agent: Optional[str] = None
    created_at: Optional[datetime] = None


# ─── DMSNotificacion ──────────────────────────────────────────────────────────

class DMSNotificacionCreate(BaseModel):
    usuario_id: int
    documento_id: Optional[int] = None
    tipo: str
    titulo: str
    mensaje: str
    leida: bool = False
    accion_url: Optional[str] = None

class DMSNotificacionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    usuario_id: int
    documento_id: Optional[int] = None
    tipo: str
    titulo: str
    mensaje: str
    leida: bool
    accion_url: Optional[str] = None
    created_at: Optional[datetime] = None


# ─── DMSKPIDiario ─────────────────────────────────────────────────────────────

class DMSKPIDiarioResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    fecha: datetime
    total_documentos: int
    documentos_creados: int
    documentos_aprobados: int
    documentos_vencidos: int
    firmas_pendientes: int
    firmas_completadas: int
    workflows_activos: int
    tamanio_total_mb: Optional[Decimal] = None
    created_at: Optional[datetime] = None


# ─── Dashboard KPIs ───────────────────────────────────────────────────────────

class DMSDashboardKPIs(BaseModel):
    """KPIs agregados para el dashboard del módulo DMS."""
    total_documentos: int = 0
    documentos_activos: int = 0
    documentos_vencidos: int = 0
    documentos_proximos_vencer: int = 0
    firmas_pendientes: int = 0
    workflows_activos: int = 0
    expedientes_activos: int = 0
    cumplimiento_pct: float = 0.0
    categorias_total: int = 0
    versiones_hoy: int = 0
    tamanio_total_mb: float = 0.0
    alertas_criticas: int = 0
    documentos_creados_hoy: int = 0
    aprobaciones_pendientes: int = 0

File: frontend/src/api/dms.ts
/**
 * La gestión documental contra el servidor.
 *
 * De las catorce pantallas del módulo, siete no consultaban nada y las otras
 * siete solo leían: NINGUNA tenía una sola mutación. El servidor, en cambio,
 * expone sesenta y seis rutas con altas, cambios y bajas completas en quince
 * recursos. Este archivo es lo que faltaba entre las dos cosas.
 *
 * Los tipos siguen a los `Response` de `backend/app/application/schemas/dms.py`.
 */
import { apiClient } from './client'

export interface Categoria {
  id: number; nombre: string; descripcion?: string | null
  codigo?: string | null; icono?: string | null; color?: string | null
  activo: boolean; created_at?: string | null
}

export interface Carpeta {
  id: number; nombre: string; descripcion?: string | null
  padre_id?: number | null; ruta?: string | null
  icono?: string | null; color?: string | null
  es_publica: boolean; creado_por_id?: number | null
  created_at?: string | null; updated_at?: string | null
}

export interface TipoDocumento {
  id: number; nombre: string; descripcion?: string | null
  categoria_id?: number | null; codigo?: string | null
  extensiones_permitidas?: string | null
  requiere_firma: boolean; requiere_aprobacion: boolean
  dias_vigencia?: number | null; activo: boolean
}

export interface CampoMetadato {
  id: number; tipo_documento_id: number; nombre: string; etiqueta: string
  tipo_dato: string; opciones?: string | null
  requerido: boolean; orden: number
}

export interface Documento {
  id: number; codigo?: string | null; nombre: string
  descripcion?: string | null
  tipo_documento_id?: number | null; carpeta_id?: number | null
  estado: string; version_actual: string; version_numero: number
  tags?: string | null
  es_confidencial: boolean; permite_descarga: boolean; permite_impresion: boolean
  fecha_vigencia_inicio?: string | null; fecha_vigencia_fin?: string | null
  propietario_id?: number | null
  hcm_colaborador_id?: number | null; hcm_conductor_id?: number | null
  tms_vehiculo_id?: number | null
  modulo_origen?: string | null; referencia_externa_id?: number | null
  created_at?: string | null; updated_at?: string | null
}

/** Lo que devuelve el LISTADO: menos campos, pero con los nombres ya resueltos. */
export interface DocumentoEnLista {
  id: number; codigo?: string | null; nombre: string
  tipo_nombre?: string | null; estado: string; version_actual: string
  fecha_vigencia_inicio?: string | null; fecha_vigencia_fin?: string | null
  propietario_nombre?: string | null; created_at?: string | null
}

export interface Version {
  id: number; documento_id: number
  numero_version: string; version_numero: number; es_mayor: boolean
  nombre_archivo?: string | null; ruta_archivo?: string | null
  url_archivo?: string | null; tamanio_bytes?: number | null
  tipo_mime?: string | null; hash_md5?: string | null
  comentario?: string | null; ocr_texto?: string | null
  creado_por_id?: number | null; created_at?: string | null
}

export interface Firma {
  id: number; documento_id: number; version_id?: number | null
  firmante_id?: number | null; tipo_firma: string; estado: string
  fecha_firma?: string | null; ip_firma?: string | null
  dispositivo?: string | null; observaciones?: string | null
  orden?: number | null; created_at?: string | null
}

export interface Workflow {
  id: number; nombre: string; descripcion?: string | null
  tipo_documento_id?: number | null; activo: boolean
  dias_limite?: number | null; created_at?: string | null
}

export interface WorkflowPaso {
  id: number; workflow_id: number; nombre: string; tipo?: string | null
  orden: number; responsable_id?: number | null
  responsable_rol?: string | null; dias_limite?: number | null
  es_obligatorio: boolean
}

export interface Instancia {
  id: number; workflow_id: number; documento_id?: number | null
  estado: string; iniciado_por_id?: number | null; paso_actual?: number | null
  fecha_inicio?: string | null; fecha_limite?: string | null
  fecha_fin?: string | null; comentario_cierre?: string | null
}

export interface Expediente {
  id: number; codigo?: string | null; nombre: string
  tipo: string; estado: string; descripcion?: string | null
  hcm_colaborador_id?: number | null; hcm_conductor_id?: number | null
  tms_vehiculo_id?: number | null; propietario_id?: number | null
  completitud_pct: number
  created_at?: string | null; updated_at?: string | null
}

export interface ExpedienteDocumento {
  id: number; expediente_id: number; documento_id: number
  tipo_requerido?: string | null; es_obligatorio: boolean
}

export interface Retencion {
  id: number; nombre: string; tipo_documento_id?: number | null
  dias_retencion_activo?: number | null; dias_retencion_total?: number | null
  accion_vencimiento?: string | null; normativa?: string | null
  activo: boolean
}

export interface RegistroAuditoria {
  id: number; documento_id?: number | null; version_id?: number | null
  usuario_id?: number | null; accion: string; detalle?: string | null
  ip_origen?: string | null; user_agent?: string | null
  created_at?: string | null
}

export interface Notificacion {
  id: number; usuario_id?: number | null; documento_id?: number | null
  tipo?: string | null; titulo: string; mensaje?: string | null
  leida: boolean; accion_url?: string | null; created_at?: string | null
}

export interface TableroDMS {
  total_documentos: number; documentos_activos: number
  documentos_vencidos: number; documentos_proximos_vencer: number
  firmas_pendientes: number; workflows_activos: number
  expedientes_activos: number; cumplimiento_pct: number
  categorias_total: number; versiones_hoy: number
  tamanio_total_mb: number; alertas_criticas: number
  documentos_creados_hoy: number; aprobaciones_pendientes: number
}

const get = <T,>(ruta: string, params?: Record<string, unknown>) =>
  apiClient.get<T>(ruta, { params }).then(r => r.data)

export const dmsApi = {
  tablero: () => get<TableroDMS>('/dms/dashboard/kpis'),

  categorias: () => get<Categoria[]>('/dms/categorias'),
  crearCategoria: (d: Partial<Categoria>) =>
    apiClient.post<Categoria>('/dms/categorias', d).then(r => r.data),
  editarCategoria: (id: number, d: Partial<Categoria>) =>
    apiClient.put<Categoria>(`/dms/categorias/${id}`, d).then(r => r.data),
  borrarCategoria: (id: number) => apiClient.delete(`/dms/categorias/${id}`),

  carpetas: (padre_id?: number) => get<Carpeta[]>('/dms/carpetas', { padre_id }),
  crearCarpeta: (d: Partial<Carpeta>) =>
    apiClient.post<Carpeta>('/dms/carpetas', d).then(r => r.data),
  editarCarpeta: (id: number, d: Partial<Carpeta>) =>
    apiClient.put<Carpeta>(`/dms/carpetas/${id}`, d).then(r => r.data),
  borrarCarpeta: (id: number) => apiClient.delete(`/dms/carpetas/${id}`),

  tipos: () => get<TipoDocumento[]>('/dms/tipos-documento'),
  crearTipo: (d: Partial<TipoDocumento>) =>
    apiClient.post<TipoDocumento>('/dms/tipos-documento', d).then(r => r.data),
  editarTipo: (id: number, d: Partial<TipoDocumento>) =>
    apiClient.put<TipoDocumento>(`/dms/tipos-documento/${id}`, d).then(r => r.data),
  camposDeTipo: (id: number) => get<CampoMetadato[]>(`/dms/tipos-documento/${id}/campos`),
  crearCampo: (d: Partial<CampoMetadato>) =>
    apiClient.post<CampoMetadato>('/dms/campos-metadato', d).then(r => r.data),
  editarCampo: (id: number, d: Partial<CampoMetadato>) =>
    apiClient.put<CampoMetadato>(`/dms/campos-metadato/${id}`, d).then(r => r.data),
  borrarCampo: (id: number) => apiClient.delete(`/dms/campos-metadato/${id}`),

  // Pagina con `page`/`per_page`, no con `skip`/`limit`: es lo que espera el
  // servidor, y mandarle los otros nombres los ignora en silencio y devuelve
  // siempre la primera página de veinte.
  documentos: (f?: {
    carpeta_id?: number; tipo_documento_id?: number; estado?: string
    q?: string; page?: number; per_page?: number
  }) => get<DocumentoEnLista[]>('/dms/documentos', { per_page: 200, ...f }),
  documento: (id: number) => get<Documento>(`/dms/documentos/${id}`),
  crearDocumento: (d: Partial<Documento>) =>
    apiClient.post<Documento>('/dms/documentos', d).then(r => r.data),
  editarDocumento: (id: number, d: Partial<Documento>) =>
    apiClient.put<Documento>(`/dms/documentos/${id}`, d).then(r => r.data),
  borrarDocumento: (id: number) => apiClient.delete(`/dms/documentos/${id}`),
  cambiarEstado: (id: number, estado: string, comentario?: string) =>
    apiClient.put(`/dms/documentos/${id}/estado`,
      { estado, comentario }).then(r => r.data),
  versiones: (id: number) => get<Version[]>(`/dms/documentos/${id}/versiones`),
  metadatos: (id: number) => get<any[]>(`/dms/documentos/${id}/metadatos`),

  firmas: (f?: { documento_id?: number; estado?: string; firmante_id?: number }) =>
    get<Firma[]>('/dms/firmas', { limit: 200, ...f }),
  crearFirma: (d: Partial<Firma>) =>
    apiClient.post<Firma>('/dms/firmas', d).then(r => r.data),
  firmar: (id: number, observaciones?: string) =>
    apiClient.put<Firma>(`/dms/firmas/${id}/firmar`, { observaciones }).then(r => r.data),
  rechazarFirma: (id: number, observaciones?: string) =>
    apiClient.put<Firma>(`/dms/firmas/${id}/rechazar`, { observaciones }).then(r => r.data),

  workflows: () => get<Workflow[]>('/dms/workflows'),
  crearWorkflow: (d: Partial<Workflow>) =>
    apiClient.post<Workflow>('/dms/workflows', d).then(r => r.data),
  editarWorkflow: (id: number, d: Partial<Workflow>) =>
    apiClient.put<Workflow>(`/dms/workflows/${id}`, d).then(r => r.data),
  borrarWorkflow: (id: number) => apiClient.delete(`/dms/workflows/${id}`),
  pasos: (id: number) => get<WorkflowPaso[]>(`/dms/workflows/${id}/pasos`),
  crearPaso: (d: Partial<WorkflowPaso>) =>
    apiClient.post<WorkflowPaso>('/dms/workflows/pasos', d).then(r => r.data),
  editarPaso: (id: number, d: Partial<WorkflowPaso>) =>
    apiClient.put<WorkflowPaso>(`/dms/workflows/pasos/${id}`, d).then(r => r.data),
  borrarPaso: (id: number) => apiClient.delete(`/dms/workflows/pasos/${id}`),

  instancias: (f?: { estado?: string; documento_id?: number }) =>
    get<Instancia[]>('/dms/instancias', { limit: 200, ...f }),
  crearInstancia: (d: Partial<Instancia>) =>
    apiClient.post<Instancia>('/dms/instancias', d).then(r => r.data),
  avanzarInstancia: (id: number, comentario?: string) =>
    apiClient.put<Instancia>(`/dms/instancias/${id}/avanzar`, { comentario }).then(r => r.data),
  cancelarInstancia: (id: number, comentario?: string) =>
    apiClient.put<Instancia>(`/dms/instancias/${id}/cancelar`, { comentario }).then(r => r.data),

  expedientes: (f?: { tipo?: string; estado?: string }) =>
    get<Expediente[]>('/dms/expedientes', { limit: 200, ...f }),
  expediente: (id: number) => get<Expediente>(`/dms/expedientes/${id}`),
  documentosDeExpediente: (id: number) =>
    get<ExpedienteDocumento[]>(`/dms/expedientes/${id}/documentos`),
  crearExpediente: (d: Partial<Expediente>) =>
    apiClient.post<Expediente>('/dms/expedientes', d).then(r => r.data),
  editarExpediente: (id: number, d: Partial<Expediente>) =>
    apiClient.put<Expediente>(`/dms/expedientes/${id}`, d).then(r => r.data),
  borrarExpediente: (id: number) => apiClient.delete(`/dms/expedientes/${id}`),
  adjuntarAExpediente: (d: Partial<ExpedienteDocumento>) =>
    apiClient.post<ExpedienteDocumento>('/dms/expedientes/documentos', d).then(r => r.data),
  quitarDeExpediente: (id: number) =>
    apiClient.delete(`/dms/expedientes/documentos/${id}`),

  retenciones: () => get<Retencion[]>('/dms/retencion'),
  // El servidor mira una ventana fija de treinta días; no recibe parámetro.
  vencimientos: () => get<any>('/dms/retencion/vencimientos'),
  crearRetencion: (d: Partial<Retencion>) =>
    apiClient.post<Retencion>('/dms/retencion', d).then(r => r.data),
  editarRetencion: (id: number, d: Partial<Retencion>) =>
    apiClient.put<Retencion>(`/dms/retencion/${id}`, d).then(r => r.data),

  auditoria: (f?: { documento_id?: number; usuario_id?: number; accion?: string; limit?: number }) =>
    get<RegistroAuditoria[]>('/dms/auditoria', { limit: 200, ...f }),

  notificaciones: (leida?: boolean) =>
    get<Notificacion[]>('/dms/notificaciones', { leida }),
  marcarLeida: (id: number) =>
    apiClient.put<Notificacion>(`/dms/notificaciones/${id}/leer`).then(r => r.data),
}

/** «2,4 MB». Los bytes crudos no le dicen nada a nadie en una tabla. */
export function peso(bytes?: number | null): string {
  const b = bytes ?? 0
  if (!b) return '—'
  if (b >= 1_073_741_824) return `${(b / 1_073_741_824).toFixed(1)} GB`
  if (b >= 1_048_576) return `${(b / 1_048_576).toFixed(1)} MB`
  if (b >= 1024) return `${Math.round(b / 1024)} KB`
  return `${b} B`
}
